pyramid_tm/__init__.py

```diff
@@ -27,102 +35,111 @@ def default_commit_veto(request, response):
-                try:
-                    t.note(text_(request.path_info))
-                except UnicodeDecodeError:
-                    t.note(text_("Unable to decode path as unicode"))
-                response = handler(request)
-                if manager.isDoomed():
-                    raise AbortResponse(response)
-                if commit_veto is not None:
-                    veto = commit_veto(request, response)
-                    if veto:
-                        raise AbortResponse(response)
-                manager.commit()
-                del request.environ['tm.active']
-                return response
-            except AbortResponse as e:
+                    t.user = text_(userid)
+            try:
+                t.note(text_(request.path_info))
+            except UnicodeDecodeError:
+                t.note(text_("Unable to decode path as unicode"))
+
+            response = handler(request)
+            if manager.isDoomed():
+                raise AbortWithResponse(response)
+
+            # check for a squashed exception and handle it
+            # this would happen if an exception view was invoked and
+            # rendered an error response
+            exc_info = getattr(request, 'exc_info', None)
+            if exc_info is not None:
+                maybe_tag_retryable(request, exc_info)
+                raise AbortWithResponse(response)
+
+            if commit_veto is not None:
+                veto = commit_veto(request, response)
+                if veto:
+                    raise AbortWithResponse(response)
+            manager.commit()
+            return response
+
+        except AbortWithResponse as e:
+            manager.abort()
+            return e.response
+
+        except Exception:
+            exc_info = sys.exc_info()
+            try:
+                maybe_tag_retryable(request, exc_info)
+                reraise(*exc_info)
+
+            finally:
+                # keep the manager alive until after invoke_exception_view
                 manager.abort()
-                del request.environ['tm.active']
-                return e.response
-            except:
-                exc_info = sys.exc_info()
-                try:
-                    retryable = manager._retryable(*exc_info[:-1])
-                    manager.abort()
-                    if (number <= 0) or (not retryable):
-                        del request.environ['tm.active']
-                        reraise(*exc_info)
-                finally:
-                    del exc_info # avoid leak
+
+                del exc_info # avoid leak
+
+        # cleanup any changes we made to the request
+        finally:
+            del environ['tm.active']
+            del environ['tm.manager']
 
     return tm_tween
 
 
+def maybe_tag_retryable(request, exc_info):
+    if request.tm._retryable(*exc_info[:-1]):
+        exc = exc_info[1]
+        if exc:
+            zope.interface.directlyProvides(exc, IRetryableError)
```

Comment:
I just pushed a version that uses `z.i.alsoProvides` thanks to @davisagli catching the issue. Super subtle!